Runtime/Scripts/DataCollection/RGDataCollection.cs

```diff
@@ -106,26 +108,60 @@ public void SaveReplayDataInfo(long clientId, RGStateActionReplayData replayData
             RGDebug.LogVerbose($"DataCollection[{clientId}] - Saving replay data for tick {replayData.tickInfo.tick}");
             
             // Add the new replay data to a new or existing mapping in our client replay data dictionary
-            var replayDatas = _clientIdToReplayData.GetOrAdd(clientId, new List<RGStateActionReplayData>());
-            replayDatas.Add(replayData);
+            var replayDataTasks = _clientIdToReplayDataTasks.GetOrAdd(clientId, v => new List<Task>());
+            var validationDataTasks = _clientIdToValidationDataTasks.GetOrAdd(clientId, v => new List<Task>());
+            
+            var filePath =
+                GetSessionDirectory($"replayData/{clientId}/rgbot_replay_data_{replayData.tickInfo.tick}.txt");
+            var task = File.WriteAllTextAsync(filePath, replayData.ToSerialized());
+            replayDataTasks.Add(task);
 
             // If the replay data has a validation, queue up a screenshot
             if (replayData.validationResults?.Length > 0)
             {
                 RGDebug.LogVerbose($"DataCollection[{clientId}] - Also saving a screenshot for tick {replayData.tickInfo.tick}");
                 _screenshotTicksRequested.Enqueue(replayData.tickInfo.tick);
+
+                var validations = replayData.validationResults;
+
+                // update the validation summary data
+                var validationSummary = _clientIdToValidationSummaries.GetOrAdd(clientId, v => new RGValidationSummary(0,0,0));
+                validationSummary.passed += validations.Count(rd => rd.result == RGValidationResultType.PASS);
+                validationSummary.failed += validations.Count(rd => rd.result == RGValidationResultType.FAIL);
+                validationSummary.warnings += validations.Count(rd => rd.result == RGValidationResultType.WARNING);
+                
+                // Convert the list into JSONL format
+                List<string> jsonLines = new List<string>();
+                foreach (var validation in validations)
+                {
+                    string jsonString = JsonConvert.SerializeObject(validation);
+                    jsonLines.Add(jsonString);
+                }
+
+                // Combine the JSON strings with newline characters
+                string jsonLinesString = string.Join("\n", jsonLines) + "\n";
+                
+                //Write out the validations to the JSONL file
+                var validationFilePath =
+                    GetSessionDirectory($"validationData/{clientId}/rgbot_validations.jsonl");
+                var validationTask = File.AppendAllTextAsync(validationFilePath, jsonLinesString);
+                validationDataTasks.Add(validationTask);
```

Comment:
This doesn't provide a guarantee of ordering. If tick 1 calls `SaveReplayData` and queues a write to `rgbot_validations.jsonl` and tick 2 does the same, there's no guarantee that tick 1's write finishes before tick 2's write starts. The `AppendAllTextAsync` method opens the file and locks it exclusively, so a simultaneous call to `AppendAllTextAsync` for the same file would throw because it can't acquire a lock on the file. If the writes can complete before the next tick you'd be OK, but we shouldn't really rely on it.

If you want to ensure ordering, what you need is a "Task Queue". Basically, you build up a chain of `.ContinueWith` calls that ensure that task `N` doesn't **start** until task `N-1` completes . It's written something like this:

```csharp
private object _lock = new object();
private Task _activeTask = Task.CompletedTask;

public void QueueSomething()
{
    async Task DoTheWork()
    {
        // This actually does the work.
        // Because of the queue logic, it's guaranteed to complete fully before it gets called again.
    }

    // If you're confident QueueSomething can't be called simultaneously, you can omit the lock
    // Problems arise if two threads are messing with _activeTask at once.
    lock(_lock)
    {
        // Note that we _replace_ "_activeTask"!
        // This means that _activeTask now represents the completion of _all_ queued work so far
        // So, when the next call to QueueSomething comes in, it's not going to _start_ it's work until the entire queue up to that point completes.
        _activeTask = _activeTask.ContinueWith(_ => DoTheWork());
    }
}
```

That's a rough version. It'll have a bunch of allocations of `Task`s though, so it might not be the best approach.

Another approach would be to enqueue the write request with all the validations to a static `ConcurrentQueue` somewhere and have a single background thread grab write requests in order and flush them to disk.

Reply:
Another solution would be to buffer the validations in separate files for each tick. Then you'd end up in the same place as replay data, which doesn't have conflicting writes to the same file, but does still produce a lot of async work to be tracked and waited for.

Reply:
eesh.. it looks like our `unity.js` version of this code probably has a similar issue

For this C# case, I'm going to go with a simple solution for now.

The code will wait for the prior write to finish on the next call.  Normally that async write will complete before the next tick comes in, so this shouldn't hold up our client processing.  If we find that we're having issues holding up the client tick processing with high tick rates or other cases we can pursue a solution more in line with your proposal above.